### ISSUE TITLE
Unique session keys between each node or shared across entire mesh?

### ISSUE THREAD

**Issue by conn** [gsliepen/tinc#161]
When data is encrypted to be sent to a node, are the session keys unique between each node or is the same session key shared across the entire subnet?

To illustrate, if I had nodes A, B, and C, and they all wanted to send data to one another, would it look like this:
```
A <-- k1 --> B
B <-- k2 --> C
C <-- k3 --> A
```

Or does it look more like this:
```
A <-- k1 --> B
B <-- k1 --> C
C <-- k1 --> A
```

My main concern is having a "bad" node eavesdrop on traffic that it's not participating in. I understand that this isn't likely in routing mode, since nodes talk directly to each other, but the model is a bit more unclear to me when in other modes such as switching.

If it helps paint a clearer picture, I'm just looking to encapsulate traffic across a LAN [in case it becomes hostile](https://www.krackattacks.com/) but I'm also concerned about the compromise of a node within the mesh.

**Comment by gsliepen** [comment]
On Wed, Nov 22, 2017 at 03:20:48AM +0000, JD wrote:

> When data is encrypted to be sent to a node, are the session keys unique between each node or is the same session key shared across the entire subnet?

Session keys are always unique between nodes. However: with tinc 1.0.*,
data is encrypted in a hop-by-hop fashion. That means that if tinc
cannot route packets between two nodes directly, it will route it via
one or more intermediate nodes, each of which will decrypt and
re-encrypt the traffic. This allows the intermediate nodes to read the
data. It might also possible for a malicious node to send false
information to other nodes and possibly get traffic rerouted via it.

In tinc 1.1, data is encrypted end-to-end. That means that even if data
is routed via an intermediate node, those intermediate nodes cannot read
the data. But, there are still ways for malicious nodes to interfere. By
default, tinc is trusting all peers, and so a peer can announce it has
any Subnet it wants, and can thus cause traffic to be redirected to it.
The StrictSubnets option can be used to prevent this from happening.
However, it is only useful in router mode.

In short, if you want to separate nodes so they cannot read each other's
traffic, don't put them in the same VPN.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>

**Comment by conn** [comment]
This is informative! Thanks for such a well-written response.

> By default, tinc is trusting all peers, and so a peer can announce it has
any Subnet it wants, and can thus cause traffic to be redirected to it.
The StrictSubnets option can be used to prevent this from happening.
However, it is only useful in router mode.

This setup would work very well in a LAN so it's more or less the use case I was looking for (the alternative being to use IPSec in transport mode).

Are there any (planned?) equivalents to `StrictSubnets` for switch/hub? It might be marginally useful to allow for MAC address pinning in a similar fashion.

**Comment by gsliepen** [comment]
On Wed, Nov 22, 2017 at 05:30:44PM +0000, JD wrote:

> > By default, tinc is trusting all peers, and so a peer can announce it has
> > any Subnet it wants, and can thus cause traffic to be redirected to it.
> > The StrictSubnets option can be used to prevent this from happening.
> > However, it is only useful in router mode.
> 
> This setup would work very well in a LAN so it's more or less the use case I was looking for (the alternative being to use IPSec in transport mode).
> 
> Are there any (planned?) equivalents to `StrictSubnets` for switch/hub? It might be marginally useful to allow for MAC address pinning in a similar fashion. 

Well, it already works for switch and hub modes. And you can have MAC
address Subnets, so if you know which nodes use which MAC addresses, you
can indeed use it!

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>

**Comment by conn** [comment]
...wait really?? Since what version? You can use `Subnet = 00:16:3e:12:34:56` ?

**Comment by gsliepen** [comment]
On Wed, Nov 22, 2017 at 05:42:20PM +0000, JD wrote:

> ...wait really?? Since what version? You can use `Subnet = 00:16:3e:12:34:56` ?

Yes, since 1.0.11 (maybe even earlier). Indeed, you just write Subnet =
01:23:45:67:89:ab.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>

**Comment by conn** [comment]
That's perfect. Thanks for the information.
